apps/labs/posts/free-threaded-gc-3-14.md

```diff
@@ -0,0 +1,323 @@
+and the interpreter's call stack.
+
+By marking all objects reachable from these _known_ roots, we can identify a
+large set of objects that are definitely alive and exclude them from the more
+expensive cycle-finding part of the GC process. Since the fraction of these
+objects is typically very high, this initial marking phase provides a
+significant performance win.
+
+As a proof-of-concept, I implemented a prototype version of this optimization
+for the default build of Python 3.12. The implementation was relatively simple
+and yielded some significant performance benefit in terms of reducing GC
+collection time. Mark Shannon independently implemented the same kind of
+optimization for the default build of Python 3.14a3. Mark's optimization gave
+an overall speedup on the [pyperformance
+suite](https://pyperformance.readthedocs.io/) of 3% and the most GC heavy
+benchmark of that suite got a 50% speedup.
+
+I implemented a version of this "mark-alive" pass for the free-threaded GC as
+well. See the [GitHub PR for
+details](https://github.com/python/cpython/pull/128808). For the pyperformance
+suite, it shows an overall speedup of 0.7% and a 35% speedup on the most GC
+heavy benchmark. Additional benchmark results will be shown later in this
+post.
+
+## Optimization 2: Using Software Pre-fetch
+
+Modern CPUs are incredibly fast, but fetching data from main memory is a major
+bottleneck. A CPU first-level cache hit might take half a nanosecond, while a
+trip to main memory can take 100 nanoseconds—200 times slower. CPUs try to
+predict which memory you'll need next (hardware pre-fetching), but this fails
+with unpredictable memory access patterns, causing the CPU to stall.
+
+Inspired by a similar change in the [OCaml
+GC](https://github.com/ocaml/ocaml/pull/10195), we introduced software
+pre-fetching to the mark-alive phase of the free-threaded GC. If we know we're
+going to need an object's data soon, we can issue an instruction to the CPU to
+start loading it from memory ahead of time.
+
+In our implementation, instead of immediately accessing an object pointer, we
+push it into a special FIFO queue called a "pre-fetch buffer." When the pointer
+is pushed, we also issue a pre-fetch instruction. We pop pointers from the
+other end of the buffer and only then access the object's data. The buffer's
+size creates a time window that, when tuned correctly, hides the memory access
+latency.
+
+In order to tune the size of the pre-fetch buffer, I created an instrumented
+version of the GC that logged pre-fetch buffer operations along with object
+access timings. An example histogram produced by this is shown below. More
+details about the GC software pre-fetch are contained in the CPython
+[internal
+documentation](https://github.com/python/cpython/blob/main/InternalDocs/garbage_collector.md#software-prefetch-hinting).
+
+  <figure style={{ textAlign: 'center' }}>
+    <img 
+      src="/posts/free-threaded-gc-3-14/prefetch-hist.png"
```

Comment:
The physicist in me really wants units on the vertical axis:)

Reply:
The bins are also a little unusual. Any reason why it's multiples of 25.6 ns, and 13k steps on the x-axis?

Reply:
If it's hard to change it's not the end of the world, but if it's easy then it may be nice to polish a bit.